neqo-http3/src/push_controller.rs

```diff
@@ -25,15 +170,324 @@ impl Display for PushController {
+    ) -> Res<()> {
+        qtrace!("Cancel push_id={}", push_id);
+
+        self.check_push_id(push_id)
+            .map_err(|_| Error::InvalidStreamId)?;
+
+        match self.push_streams.get(push_id) {
+            None => {
+                qtrace!("Push has already been closed.");
+                // If we have some events for the push_id in the event queue, the caller still does not
+                // not know that the push has been closed. Otherwise return InvalidStreamId.
+                if self.conn_events.has_push(push_id) {
+                    self.conn_events.remove_events_for_push_id(push_id);
+                    Ok(())
+                } else {
+                    Err(Error::InvalidStreamId)
+                }
+            }
+            Some(PushState::PushPromise { .. }) => {
+                self.conn_events.remove_events_for_push_id(push_id);
+                base_handler.queue_control_frame(&HFrame::CancelPush { push_id });
+                self.push_streams.close(push_id);
+                Ok(())
+            }
+            Some(PushState::Active { stream_id, .. }) => {
+                self.conn_events.remove_events_for_push_id(push_id);
+                // Cancel the stream. the transport steam may already be done, so ignore an error.
+                let _ =
+                    base_handler.stream_reset(conn, *stream_id, Error::HttpRequestCancelled.code());
+                self.push_streams.close(push_id);
+                Ok(())
+            }
+            _ => Err(Error::InvalidStreamId),
+        }
+    }
+
+    pub fn push_stream_reset(&mut self, push_id: u64) {
+        qtrace!("Push stream has been reset, push_id={}", push_id);
+
+        if let Some(push_state) = self.push_streams.get(push_id) {
+            match push_state {
+                PushState::OnlyPushStream { .. } => {
+                    self.push_streams.close(push_id);
+                }
+                PushState::Active { .. } => {
+                    self.push_streams.close(push_id);
+                    self.conn_events.remove_events_for_push_id(push_id);
+                    self.conn_events.push_canceled(push_id);
+                }
+                _ => {
+                    debug_assert!(
+                        false,
+                        "Reset cannot actually happen because we do not have a stream."
+                    );
+                }
+            }
+        }
+    }
+
+    pub fn get_active_stream_id(&mut self, push_id: u64) -> Option<u64> {
+        match self.push_streams.get(push_id) {
+            Some(PushState::Active { stream_id, .. }) => Some(*stream_id),
+            _ => None,
+        }
+    }
+
+    pub fn maybe_send_max_push_id_frame(&mut self, base_handler: &mut Http3Connection) {
+        let push_done = self.push_streams.number_done();
+        if self.max_concurent_push > 0
+            && (self.current_max_push_id - push_done) <= (self.max_concurent_push / 2)
```

Comment:
I think that we can probably be a lot more aggressive about sending this one.  In the extreme, it would be OK to send an update for every push completion.  But you could set a value that is min(self.max_concurrent_push, SMALL_NUMBER_LIKE_3) if you wanted to limit the rate a little.